app/cmd/start.go

```diff
@@ -146,89 +154,176 @@ func start(c *cli.Context) (err error) {
+		}()
+
 		if err := proxyGRPCServer.Serve(proxyGRPCListener); err != nil {
 			logrus.WithError(err).Error("Stopping proxy gRPC server")
 		}
-		// graceful shutdown before exit
-		close(shutdownCh)
 	}()
 	logrus.Infof("Instance Manager proxy gRPC server listening to %v", proxyServiceAddress)
 
 	// Start process manager server
-	pm, pmGRPCServer, pmGRPCListener, err := setupProcessManagerGRPCServer(processPortRange, logsDir, processManagerServiceAddress, shutdownCh)
+	pm, pmGRPCServer, pmGRPCListener, err := setupProcessManagerGRPCServer(processPortRange, logsDir, processManagerServiceAddress, shutdown.ch)
 	if err != nil {
 		return err
 	}
 	go func() {
+		defer func() {
+			shutdown.once.Do(func() {
+				close(shutdown.ch)
+			})
+			wg.Done()
+			logrus.Info("Stopped process manager gRPC server")
+		}()
+
 		if err := pmGRPCServer.Serve(pmGRPCListener); err != nil {
 			logrus.WithError(err).Error("Stopping process manager gRPC server")
 		}
-		// graceful shutdown before exit
 		cleanup(pm)
-		close(shutdownCh)
 	}()
 	logrus.Infof("Instance Manager process manager gRPC server listening to %v", listen)
 
 	// Start SPDK server
+	var spdkGRPCServer *grpc.Server
+	var spdkGRPCListener net.Listener
 	if spdkEnabled {
-		spdkGRPCServer, spdkGRPCListener, err := setupSPDKGRPCServer(spdkPortRange, spdkServiceAddress, shutdownCh)
+		spdkGRPCServer, spdkGRPCListener, err = setupSPDKGRPCServer(spdkPortRange, spdkServiceAddress, shutdown.ch)
 		if err != nil {
 			return err
 		}
 		go func() {
+			defer func() {
+				shutdown.once.Do(func() {
+					close(shutdown.ch)
+				})
+				wg.Done()
+				logrus.Info("Stopped SPDK gRPC server")
+			}()
+
 			if err := spdkGRPCServer.Serve(spdkGRPCListener); err != nil {
 				logrus.WithError(err).Error("Stopping SPDK gRPC server")
 			}
-			// graceful shutdown before exit
-			close(shutdownCh)
 		}()
 		logrus.Infof("Instance Manager SPDK gRPC server listening to %v", spdkServiceAddress)
 	}
 
+	// Register signal handler
 	sigs := make(chan os.Signal, 1)
 	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+
 	go func() {
 		sig := <-sigs
+
 		logrus.Infof("Instance Manager received %v to exit", sig)
+		instanceGRPCServer.Stop()
+		if spdkGRPCServer != nil {
+			spdkGRPCServer.Stop()
+		}
 		pmGRPCServer.Stop()
+		proxyGRPCServer.Stop()
+		diskGRPCServer.Stop()
 	}()
 
-	return <-shutdownCh
+	wg.Wait()
+
+	if spdkEnabled {
+		stopSPDKTgt(30 * time.Second)
+	}
+
+	return nil
+}
+
+func stopSPDKTgt(timeout time.Duration) {
+	logrus.Infof("Stopping spgk_tgt process")
+
+	process, err := util.FindProcessByCmdline("spdk_tgt")
+	if err != nil {
+		logrus.WithError(err).Error("Failed to find spdk_tgt process")
+		return
+	}
+
+	logrus.Infof("Stopping spdk_tgt process")
+	if err := process.Signal(syscall.SIGTERM); err != nil {
+		logrus.WithError(err).Errorf("Failed to send SIGTERM to spdk_tgt process")
+	} else {
+		done := make(chan error, 1)
+		go func() {
+			_, err := process.Wait()
+			done <- err
```

Comment:
Should we close the done channel at some point?

Reply:
Good catch. Thank you.